code/validation_json_schemas.py

```diff
@@ -0,0 +1,190 @@
+# TODO validate type is in valid set of values for all below
+
+fv_schema = {
+    "type": "object",
+    "properties": {
+        "rate": {
+            "type": "number"
+        },
+        "nper": {
+            "type": "number"
+        },
+        "pmt": {
+            "type": "number"
+        },
+        "pv": {
+            "type": "number"
+        },
+        "type": {
+            "type": "number"
+        }
+    },
+    "required": ["rate", "nper", "pmt", "pv"]
+}
+
+pv_schema = {
+    "type": "object",
+    "properties": {
+        "rate": {
+            "type": "number"
+        },
+        "nper": {
+            "type": "number"
+        },
+        "pmt": {
+            "type": "number"
+        },
+        "fv": {
+            "type": "number"
+        },
+        "type": {
+            "type": "number"
+        }
+    },
+    "required": ["rate", "nper", "pmt", "fv"]
+}
+
+npv_schema = {
+    "type": "object",
+    "properties": {
+        "rate": {
+            "type": "number"
+        },
+        "values": {
+            "type": "array",
+            "items": {
+                "type": "number",
+                "minItems": 1
+            }
+        }
+    },
+    "required": ["rate", "values"]
+}
+
+pmt_schema = {
+    "type": "object",
+    "properties": {
+        "rate": {
+            "type": "number"
+        },
+        "nper": {
+            "type": "number"
+        },
+        "pv": {
+            "type": "number"
+        },
+        "fv": {
+            "type": "number"
+        },
+        "type": {
+            "type": "number"
+        }
+    },
+    "required": ["rate", "nper", "pv"]
+}
+
+ppmt_schema = {
+    "type": "object",
+    "properties": {
+        "rate": {
+            "type": "number"
+        },
+        "per": {
+            "type": "number"
+        },
+        "nper": {
+            "type": "number"
+        },
+        "pv": {
+            "type": "number"
+        },
+        "fv": {
+            "type": "number"
+        },
+        "type": {
+            "type": "number"
+        }
+    },
+    "required": ["rate", "per", "nper", "pv"]
+}
+
+irr_schema = {
+    "type": "object",
+    "properties": {
+        "values": {
+            "type": "array",
+            "items": {
+                "type": "number",
+                "minItems": 1
+            }
+        }
+    },
+    "required": ["values"]
+}
+
+mirr_schema = {
+    "type": "object",
+    "properties": {
+        "values": {
+            "type": "array",
+            "items": {
+                "type": "number",
+                "minItems": 1
```

Comment:
According to Excel docs, you need at least two values (one positive, one negative). numpy docs say the same thing or else it'll return nan.

Reply:
Good catch